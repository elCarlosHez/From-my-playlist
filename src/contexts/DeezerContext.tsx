import React, { useCallback, useState } from 'react';

interface IDeezerContext {
  token: string;
  setToken: React.Dispatch<React.SetStateAction<string>>;
  fetchDeezer: (url: string) => Promise<any>;
}

export const useDeezerContext = (): IDeezerContext => {
  const [token, setToken] = useState('');

  const fetchDeezer = useCallback(async (url:string): Promise<any> => {
    try {
      // Consider Dezzer has a CORS block. This method won't work in localhost
      const promise = await fetch(`${url}?access_token=${token}`);
      return await promise.json();
    } catch (error) {
      // eslint-disable-next-line no-console
      console.log('Deezer fetch Error: ', error);
      return null;
    }
  }, [token]);

  return { token, setToken, fetchDeezer };
};
