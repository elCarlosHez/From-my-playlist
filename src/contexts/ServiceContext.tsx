import { useState } from "react";
import { Playlist } from "../types/Playlist";
import { ServicesList } from "../types/ServicesList";

export interface ReturnServiceContext {
  service: ServicesList | null;
  setService: React.Dispatch<React.SetStateAction<ServicesList | null>>
  playlist: Playlist | null;  
  setPlaylist: React.Dispatch<React.SetStateAction<Playlist | null>>;
}

export const serviceContext = (): ReturnServiceContext => {
  const [service, setService] = useState<ServicesList | null>(null);
  const [playlist, setPlaylist] = useState<Playlist | null>(null);

  return {
    service,
    setService,
    playlist,
    setPlaylist
  }
};
