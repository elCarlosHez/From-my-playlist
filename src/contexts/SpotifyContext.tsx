import React, { createContext, useContext, useState } from 'react';

interface ISpotifyContext {
  token: string;
  setToken: (value: string) => void;
  fetchSpotify: <T>(url: string) => Promise<T | null>;
}

const SpotifyContext = createContext<ISpotifyContext | undefined>(undefined);

interface ISpotifyProvider {
  children: JSX.Element;
}

export const SpotifyProvider = (props: ISpotifyProvider): JSX.Element => {
  const { children } = props;
  const [token, setToken] = useState('');

  async function fetchSpotify <T>(url: string): Promise<T | null> {
    try {
      const promise = await fetch(url, {
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      });
      return await promise.json();
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error);
      return null;
    }
  }

  return (
    <SpotifyContext.Provider
      // eslint-disable-next-line react/jsx-no-constructed-context-values
      value={{
        token,
        setToken,
        fetchSpotify,
      }}
    >
      {children}
    </SpotifyContext.Provider>
  );
};

export const useSpotifyAuth = () => {
  const context = useContext(SpotifyContext);
  if (context === undefined) {
    throw new Error('useSpotifyAuth must be used within a SpotifyProvider');
  }

  return context;
};
