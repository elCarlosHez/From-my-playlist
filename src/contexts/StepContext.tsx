import { useState } from "react";

export enum Steps {
  start,
  destiny,
  transfer,
  finish
}

export interface ReturnStepContext {
  step: Steps;
  setStep: React.Dispatch<React.SetStateAction<Steps>>;
}

export const stepContext = (): ReturnStepContext => {
  const [step, setStep] = useState<Steps>(Steps.start);

  return {
    step,
    setStep,
  }
};
