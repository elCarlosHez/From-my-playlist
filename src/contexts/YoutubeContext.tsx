import React, { useCallback, useState } from "react";

interface ISpotifyProvider {
  token: string;
  setToken: React.Dispatch<React.SetStateAction<string>>;
  fetchYoutube: (url: string) => Promise<any>;
}

export const useYoutubeContext = (): ISpotifyProvider => {
  const [token, setToken] = useState('');

  const fetchYoutube = useCallback(
    async (url: string): Promise<any> => {
      try {
        const promise = await fetch(`${url}?part=snippet&mine=true`, {
          headers: {
            Accept: "application/json",
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
        });
        return await promise.json();
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error(error);
        return null;
      }
    },
    [token]
  );

  return { token, setToken, fetchYoutube };
};
