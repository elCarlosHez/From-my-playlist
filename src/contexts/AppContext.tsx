import React, { createContext, useContext, useState } from "react";
import { useDeezerContext } from "./DeezerContext";
import { useSpotifyContext } from "./SpotifyContext";
import { useYoutubeContext } from "./YoutubeContext";

interface IAppContext {
  // Spotify Context
  spotifyToken: string;
  setSpotifyToken: React.Dispatch<React.SetStateAction<string>>;
  fetchSpotify: (url: string) => Promise<any>;
  // Youtube Music Context
  youtubeToken: string;
  setYoutubeToken: React.Dispatch<React.SetStateAction<string>>;
  fetchYoutube: (url: string) => Promise<any>;
}

const AppContext = createContext<IAppContext | undefined>(undefined);

interface IAppProvider {
  children: JSX.Element;
}

export const AppProvider = (props: IAppProvider): JSX.Element => {
  const { children } = props;
  const {
    token: spotifyToken,
    setToken: setSpotifyToken,
    fetchSpotify,
  } = useSpotifyContext();

  const {
    token: youtubeToken,
    setToken: setYoutubeToken,
    fetchYoutube,
  } = useYoutubeContext();

  return (
    <AppContext.Provider
      value={{
        spotifyToken,
        setSpotifyToken,
        fetchSpotify,

        youtubeToken,
        setYoutubeToken,
        fetchYoutube,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  const context = useContext(AppContext);
  if (context === undefined) {
    throw new Error("useAppContext must be used within a AppProvider");
  }

  return context;
};
