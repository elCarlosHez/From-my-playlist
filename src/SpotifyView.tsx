import React, { useEffect, useState } from 'react';

import logo from './assets/logo.png';
import './styles/App.css';
import { useSpotifyAuth } from './contexts/SpotifyContext';
import { generateSpotifyUrl } from './services/Spotify';
import { Playlist } from './types/Playlist';
import { PlaylistQuery } from './types/PlaylistQuery';
import { TOKEN_SPOTIFY_TYPE, TokenMessage } from './types/TokenTypes';
import OpenAuthenticationPopup from './utils/openAuthenticationPopup';

function SpotifyView(): JSX.Element {
  const { token, setToken, fetchSpotify } = useSpotifyAuth();
  const [playlists, setPlaylists] = useState<Playlist[]>([]);

  useEffect(() => {
    window.addEventListener('message', (event) => {
      // For security reasons, if we recieved a message from other origin, we cancel our handle
      if (event.origin !== import.meta.env.VITE_APP_URL as string) return;

      const message:TokenMessage = event.data;
      // We check if the type we'll recieve is the type Spotify
      if (message?.type === TOKEN_SPOTIFY_TYPE) {
        setToken(message.token);
      }
    });
    const url = generateSpotifyUrl();
    OpenAuthenticationPopup(url, 'Spotify');
  }, []);

  useEffect(() => {
    // We haven't recieved the authentication token
    if (!token.length) return;
    fetchSpotify<PlaylistQuery>('https://api.spotify.com/v1/me/playlists').then(
      (data) => {
        if (data) {
          setPlaylists(data.items);
        }
      },
    );
  }, [token]);

  return (
    <>
      <header className="flex justify-center items-center flex-col">
        <img alt="From My Playlist" className="max-w-full mb-5" src={logo} />
      </header>
      <main className="grid grid-cols-12">
        <section className="col-start-6 col-span-2 flex flex-col">
          <h2 className="text-center mb-4 text-xl font-medium">
            Select your playlist
          </h2>
          {playlists.map((playlist) => (
            <button type="button" key={playlist.id} className="btn-primary mb-4">
              {playlist.name}
            </button>
          ))}
        </section>
      </main>
    </>
  );
}

export default SpotifyView;
