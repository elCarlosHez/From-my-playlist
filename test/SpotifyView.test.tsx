import React from "react";
import renderer from "react-test-renderer";
import SpotifyView from "../src/SpotifyView";
import { SpotifyProvider } from "../src/contexts/SpotifyContext";
import { BrowserRouter } from "react-router-dom";

// @ts-ignore
global.window = Object.create(window);
const url = "http://dummy.com";
Object.defineProperty(window, "location", {
    value: {
       href: url
    },
    writable: true
});

Object.defineProperty(window, "open", {
  value: () => {},
  writable: true
});

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();
  expect(result).toBeDefined();
  return result as renderer.ReactTestRendererJSON;
}

describe("SpotifyView tests", () => {
  it("Renders without crash", () => {
    const component = renderer.create(
      <BrowserRouter>
        <SpotifyProvider>
          <SpotifyView />
        </SpotifyProvider>
      </BrowserRouter>
    );
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
