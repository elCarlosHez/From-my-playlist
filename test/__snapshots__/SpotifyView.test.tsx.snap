// Vitest Snapshot v1

exports[`SpotifyView tests > Renders without crash 1`] = `
[
  <header
    className="flex justify-center items-center flex-col"
  >
    <img
      alt="From My Playlist"
      className="max-w-full mb-5"
      src="/src/assets/logo.png"
    />
  </header>,
  <main
    className="grid grid-cols-12"
  >
    <section
      className="col-start-6 col-span-2 flex flex-col"
    >
      <h2
        className="text-center mb-4 text-xl font-medium"
      >
        Select your playlist
      </h2>
    </section>
  </main>,
]
`;
